akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala

```diff
@@ -76,14 +77,31 @@ private[akka] object RoutingLogics {
 
     private var currentRoutees: Array[ActorRef[T]] = _
 
-    override def selectRoutee(): ActorRef[T] = {
+    override def selectRoutee(msg: T): ActorRef[T] = {
       val selectedIdx = ThreadLocalRandom.current().nextInt(currentRoutees.length)
       currentRoutees(selectedIdx)
     }
+
     override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
       currentRoutees = newRoutees.toArray
     }
+  }
+
+  final class ConsistentHashingLogic[T](virtualNodesFactor: Int, mapping: RoutingHashExtractor[T])
+      extends RoutingLogic[T] {
+    require(virtualNodesFactor > 0, "virtualNodesFactor has to be a positive integer")
+
+    private var currentRoutees: Set[ActorRef[T]] = Set.empty
 
+    private var consistentHash: ConsistentHash[ActorRef[T]] = ConsistentHash(currentRoutees, virtualNodesFactor)
+
+    override def selectRoutee(msg: T): ActorRef[T] = consistentHash.nodeFor(mapping(msg))
+
+    override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
+      val withoutOld = currentRoutees.diff(newRoutees).foldLeft(consistentHash)(_ :- _)
+      consistentHash = newRoutees.diff(currentRoutees).foldLeft(withoutOld)(_ :+ _)
```

Comment:
In the end the `ConsistentHash` is using `toString` on the routees ActorRef, and that will include the address for remote ActorRef but not for local ActorRef. It must be the same hashing on all nodes, so therefore it's important to always include the address.

In the classic implementation that is done in the `ConsistentRoutee` https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/routing/ConsistentHashing.scala#L423
`path.toStringWithAddress(selfAddress)`